redis/src/aio.rs

```diff
@@ -677,32 +679,41 @@ where
         }
     }
 
-    fn send_result(self: Pin<&mut Self>, result: Result<I, E>) {
-        let self_ = self.project();
-        let response = {
+    fn send_result(mut self: Pin<&mut Self>, result: Result<I, E>) {
+        let self_ = self.as_mut().project();
+
+        {
             let entry = match self_.in_flight.front_mut() {
                 Some(entry) => entry,
                 None => return,
             };
+
             match result {
                 Ok(item) => {
                     entry.buffer.push(item);
-                    if entry.response_count > entry.buffer.len() {
-                        // Need to gather more response values
-                        return;
-                    }
-                    Ok(mem::take(&mut entry.buffer))
                 }
-                // If we fail we must respond immediately
-                Err(err) => Err(err),
+                Err(err) => {
+                    self_.redis_errors.push_back(err);
+                }
             }
-        };
+
+            if entry.response_count > (entry.buffer.len() + self_.redis_errors.len()) {
+                // Need to gather more response values
+                return;
+            }
+        }
 
         let entry = self_.in_flight.pop_front().unwrap();
-        // `Err` means that the receiver was dropped in which case it does not
-        // care about the output and we can continue by just dropping the value
-        // and sender
-        entry.output.send(response).ok();
+        let mut redis_errors = mem::take(self_.redis_errors);
+        if !redis_errors.is_empty() {
+            // Send the first redis_error found:
+            entry
+                .output
+                .send(Err(redis_errors.pop_front().unwrap()))
```

Comment:
So we just send the first error here. Should we change the API so that we can send all errors, instead? Why not?

Reply:
I agree that would be a better API, but I'm not sure we can easily accommodate without making changes to the `ConnectionLike` trait. 

If possible, I would like to get the minimal approach in this PR added because it mimics existing behavior -- i.e., returning the `Err` from the first failure in the pipe -- minus the data corruption. I think it can easily be added to a 21.x release as well.

Reply:
What if I consolidate all the errors into an aggregated `RedisError`? Not sure how much work that is offhand, but I'm sure it'd be doable.

Reply:
I've simplified it to only store the first error, so that it now matches the behavior of the other Connection types.

Reply:
Moved mutable fields onto the Inflight entry, which is more appropriate here.